src/Starcounter.Database.Extensions/NestedTransactor.cs

```diff
@@ -0,0 +1,103 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Starcounter.Database.Extensions
+{
+    public class NestedTransactor : TransactorBase
+    {
+        ThreadLocal<IDatabaseContext> _current = new ThreadLocal<IDatabaseContext>();
+
+        internal class NestedTransactionContext : ContextBase
+        {
+            public NestedTransactionContext(IDatabaseContext inner) : base(inner) { }
+        }
+
+        public NestedTransactor(ITransactor innerTransactor) : base(innerTransactor) { }
+
+        public override void Transact(Action<IDatabaseContext> action, TransactOptions options = null)
+        {
+            var nested = _current.Value;
+            if (nested != null)
+            {
+                action(new NestedTransactionContext(nested));
+                return;
+            }
+
+            base.Transact(action, options);
+        }
+
+        public override T Transact<T>(Func<IDatabaseContext, T> function, TransactOptions options = null)
+        {
+            var nested = _current.Value;
+            if (nested != null)
+            {
+                return function(new NestedTransactionContext(nested));
+            }
+
+            return base.Transact(function, options);
+        }
+
+        public override Task TransactAsync(Action<IDatabaseContext> action, TransactOptions options = null)
+        {
+            var nested = _current.Value;
+            if (nested != null)
+            {
+                action(new NestedTransactionContext(nested));
```

Comment:
Child and parent will show a different behavior In the case of exception being thrown from the action. A parent will return a failed task, while a child will throw instantly. The same applies to the next overload.

Reply:
Line numbers seem to mismatch here, probably because something has changed, so I'm not sure what 2 overloads you are referring to? I'm guessing Async versions, since you refer to tasks being returned, while line numbers seem to indicate sync ones.

WTBS, I need to think this through a bit. Because if we are talking about Async versions (which I think we do), isn't it so that the can execute on the current thread of they get the context lock, and then they will also raise immediate exception, no?

Also, I need to think through this trait in general - if it must necessarily be the exact same behavior for a nested execution path as a parent non-nested one. Probably so, but I want to give it some thinking first.

For now, please do comment on the first 2 paragraphs. Maybe include signatures of the overloads this applies to.

Reply:
It concerns `public override Task TransactAsync(Action<IDatabaseContext> action, TransactOptions options = null)` and `public override Task<T> TransactAsync<T>(Func<IDatabaseContext, T> function, TransactOptions options = null)`

Throwing an exception synchronously is legal - it just differs from default Transactor behavior. I guess we can let it be.

Reply:
PoLA here? I guess, in the light of that default behavior, it would always be the same as that. But let's say we wouldn't consider any default behavior. I guess returning a faulted task would still be the choice, right?

If so, I think we should adapt. Wouldn't it also apply to Async overloads `Task TransactAsync(Func<IDatabaseContext, Task> function, ...` and `Task<T> TransactAsync<T>(Func<IDatabaseContext, Task<T>> function, ...`?

Reply:
> But let's say we wouldn't consider any default behavior. I guess returning a faulted task would still be the choice, right?

I think returning a failed task is more common, but there is no such convention, because an async function is still just a function returning `Task`.

>If so, I think we should adapt. Wouldn't it also apply to Async overloads Task TransactAsync(Func<IDatabaseContext, Task> function, ... and Task<T> TransactAsync<T>(Func<IDatabaseContext, Task<T>> function, ...?

It also applies to them, but it's harder to write an async function which throws an exception synchronously.

Reply:
And for clients who use it with await: `await db.TransactAsync(...)` there will be no difference at all.

Reply:
> And for clients who use it with await: `await db.TransactAsync(...)` there will be no difference at all.

Yes, that much is clear.

I did some reading now, and as you say, it doesn't seem to be any certain right or wrong.

I at least think we should address this by returning a failed task, as its aligned with the default transactor, and as no more "wrong" than raising exceptions straight away.

Reply:
Done.